### PR TITLE
mailbox security enhancement & u30 vm enable

### PR DESCRIPTION
**Pull request by xuhz**
1. h/w mailbox security enhancement
   a. recv pkt rate check at pkt layer -- will disable mailbox if recv pkt rate is high
   b. recv msg type check at msg layer. -- admin set disabled channels, msg arrived on disabled channel will be discarded
   c. recv metrics exported as sysfs node
   d. introduce a bin sysfs node to write raw pkts to h/w mailbox for test purpose
2. u30 vm enablement
   a. in cloud env, if xclbin download channel is disabled, the device is also marked as ready without sw mailbox open(no msd/mpd plugin)
   b. preload xclbin support. after the reset, if xclbin is preloaded/cached, it will be reloaded again.

### CODE REVIEW

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c [hunk 1]
```diff
@@ -1210,7 +1270,21 @@ static bool do_hw_rx(struct mailbox_channel *ch)
...
+	/*
+	 * Don't trust the peer. If we think the peer is doing something
+	 * malicious, we disable interrupt and don't handle the pkts. Once
+	 * this happened, user can't use mailbox anymore before admin manually
+	 * recovers the mailbox by doing 'xbmgmt reset --hot --card xxx'
+	 * This is the protection at the pkt layer. The malicious user can
+	 * still escape the protection here by carefully control the sending
+	 * pkt rate. At msg layer, we have another type of protection -- we
+	 * discard those msg requests which are disabled by admin. eg,
+	 * xclbin download (type 0x8) is not expected to show up on h/w mailbox
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c:1282]: isn't the channel switch good enough for this? if you don't want to see 0x8 on hw channel, you can redirect it to sw channel (assuming you already have some kind of protection in place for sw channel).

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c [hunk 2]
```diff
@@ -1222,7 +1296,7 @@ static bool do_hw_rx(struct mailbox_channel *ch)
 		return true;
 	case PKT_MSG_START:
 		if (ch->mbc_cur_msg) {
-			MBX_ERR(mbx, "Received partial msg (id 0x%llx)\n",
+			MBX_DBG(mbx, "Received partial msg (id 0x%llx)\n",
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c:1299]: Why not print it out by default? if this shows up in your fuzz test too often, maybe we should consider treating it as malicious guest after seeing it xx times and turn off the mailbox?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c [hunk 3]
```diff
@@ -1233,18 +1307,18 @@ static bool do_hw_rx(struct mailbox_channel *ch)
 			pkt->body.msg_start.msg_size);
 
 		if (!ch->mbc_cur_msg) {
-			MBX_ERR(mbx, "got unexpected msg start pkt\n");
+			MBX_DBG(mbx, "got unexpected msg start pkt\n");
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c:1310]: ditto

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c [hunk 4]
```diff
@@ -1233,18 +1307,18 @@ static bool do_hw_rx(struct mailbox_channel *ch)
...
 			reset_pkt(pkt);
 		}
 		break;
 	case PKT_MSG_BODY:
 		if (!ch->mbc_cur_msg) {
-			MBX_ERR(mbx, "got unexpected msg body pkt\n");
+			MBX_DBG(mbx, "got unexpected msg body pkt\n");
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c:1316]: ditto

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c [hunk 5]
```diff
@@ -1233,18 +1307,18 @@ static bool do_hw_rx(struct mailbox_channel *ch)
...
 			reset_pkt(pkt);
 		}
 		break;
 	default:
-		MBX_ERR(mbx, "invalid mailbox pkt type: %d\n", type);
+		MBX_DBG(mbx, "invalid mailbox pkt type: %d\n", type);
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c:1321]: ditto

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_sysfs.c [hunk 6]
```diff
@@ -513,11 +524,20 @@ static ssize_t ready_show(struct device *dev,
...
 		 * If xocl and xclmgmt are not in the same daemon,
-		 * mark the card as ready only when both MB channel
-		 * and daemon are ready
+		 * mark the card as ready when
+		 *  1. both MB channel and daemon are ready
+		 *  This is for case cloud vendor controls the xclbin download,
+		 *  like azure, aws F1
+		 *  2. MB channel is ready
+		 *     and
+		 *     XCL_MAILBOX_REQ_LOAD_XCLBIN is disabled
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_sysfs.c:533]: wow... "ready" state is a very generic state for a device.
but, here, we use disable state of one specific mailbox op code to determine the generic "ready" state of the board. this sounds very hacky to me. it feels that we are missing one abstract layer here...
what if admin disabled another op code? what if this XCL_MAILBOX_REQ_LOAD_XCLBIN op code is removed completely in future release?
i'm wondering if we can introduce another flag for mgmt to explicitly tell user that there will be no daemon required?

**Review by maxzhen — CHANGES_REQUESTED** [review]

**Comment by xilinxgitops** [comment]
retest this please

**Comment by dayeh-xilinx** [comment]
retest this please

File: src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-sysfs.c [hunk 7]
```diff
@@ -218,6 +218,33 @@ static ssize_t dev_offline_show(struct device *dev,
...
+	(void) xocl_mailbox_set(lro, CHAN_DISABLE, val);
+	xclmgmt_connect_notify(lro, true);
+
+	return count;
+}
+static ssize_t disable_mailbox_channel_show(struct device *dev,
+	struct device_attribute *attr, char *buf)
+{
+	struct xclmgmt_dev *lro = dev_get_drvdata(dev);
+	uint64_t ch_switch = 0;
```
[the remark this answers is not in the document]
Comment (xuhz) [reply src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-sysfs.c]: changed

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c (hunk 2 above)
Comment (xuhz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c]: flood of err msg will be on console and makes the host slow. changed to MBX_WARN

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_sysfs.c (hunk 6 above)
Comment (xuhz) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_sysfs.c]: changed as discussed

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c (hunk 1 above)
Comment (xuhz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/mailbox.c:1282]: 2 per channel variables required

**Review by maxzhen — APPROVED** [review]

**Review by larry9523 — APPROVED** [review]

**Review by houlz0507 — APPROVED** [review]

**Comment by abalasa** [comment]
retest this please

**Comment by dayeh-xilinx** [comment]
retest this please

**Pull request merged**